src/VisualStudio/CSharp/Impl/LanguageServerClient/CSharpLanguageServerClient.cs

```diff
@@ -0,0 +1,155 @@
+        private readonly Shell.IAsyncServiceProvider _asyncServiceProvider;
+
+        /// <summary>
+        /// Gets the name of the language client (displayed to the user).
+        /// </summary>
+        public string Name { get; } = "C# Language Server Client";
+
+        /// <summary>
+        /// Gets the configuration section names for the language client. This may be null if the language client
+        /// does not provide settings.
+        /// </summary>
+        public IEnumerable<string> ConfigurationSections { get; } = null;
+
+        /// <summary>
+        /// Gets the initialization options object the client wants to send when 'initialize' message is sent.
+        /// This may be null if the client does not need custom initialization options.
+        /// </summary>
+        public object InitializationOptions { get; } = null;
+
+        /// <summary>
+        /// Gets the list of file names to watch for changes.  Changes will be sent to the server via 'workspace/didChangeWatchedFiles'
+        /// message.  The files to watch must be under the current active workspace.  The file names can be specified as a relative
+        /// paths to the exact file, or as glob patterns following the standard in .gitignore see https://www.kernel.org/pub/software/scm/git/docs/gitignore.html files.
+        /// </summary>
+        public IEnumerable<string> FilesToWatch { get; } = null;
+
+#pragma warning disable CS0067
+        public event AsyncEventHandler<EventArgs> StartAsync;
+        public event AsyncEventHandler<EventArgs> StopAsync;
+#pragma warning restore CS0067
+
+        [ImportingConstructor]
+        public CSharpLanguageServerClient(
+            PrimaryWorkspace primaryWorkspace,
+            [Import(typeof(SAsyncServiceProvider))]Shell.IAsyncServiceProvider asyncServiceProvider)
+        {
+            _primaryWorkspace = primaryWorkspace;
+            _asyncServiceProvider = asyncServiceProvider;
+        }
+
+        public async Task<Connection> ActivateAsync(CancellationToken token)
+        {
+            // Establish a connection to the WellKnownServiceHubServices.LanguageServer
+            string codeAnalysisHostGroupId = await GetHostGroupIdAsync(token).ConfigureAwait(false);
+            var hostGroup = new HostGroup(codeAnalysisHostGroupId);
+            var serviceDescriptor = new ServiceDescriptor(WellKnownServiceHubServices.LanguageServer) { HostGroup = hostGroup };
+            var client = new HubClient("ManagedLanguage.IDE.CSharpLSPServerClient");
+            var stream = await client.RequestServiceAsync(serviceDescriptor, token).ConfigureAwait(true);
+            return new Connection(stream, stream);
+        }
+
+        public async Task<string> GetHostGroupIdAsync(CancellationToken cancellationToken)
+        {
+            var client = await _primaryWorkspace.Workspace.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
@heejaechang I'm seeing a problem here: if the user attempts to search during solution load of a large solution, the remote host client isn't setup yet. What's the preferred pattern here? Can we request to start the client or await its startup?